lxd/profiles_utils.go

```diff
@@ -0,0 +1,147 @@
+package main
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/lxc/lxd/shared/api"
+)
+
+func doProfileUpdate(d *Daemon, name string, id int64, profile *api.Profile, req api.ProfilePut) Response {
+	// Sanity checks
+	err := containerValidConfig(d, req.Config, true, false)
+	if err != nil {
+		return BadRequest(err)
+	}
+
+	err = containerValidDevices(req.Devices, true, false)
+	if err != nil {
+		return BadRequest(err)
+	}
+
+	containers := getContainersWithProfile(d, name)
+
+	// Check if the root device is supposed to be changed or removed.
+	oldProfileRootDiskDevice := containerGetRootDiskDevice(profile.Devices)
+	newProfileRootDiskDevice := containerGetRootDiskDevice(req.Devices)
+	if len(containers) > 0 && oldProfileRootDiskDevice != nil && oldProfileRootDiskDevice["pool"] != "" {
+		if newProfileRootDiskDevice == nil || newProfileRootDiskDevice["pool"] == "" || (oldProfileRootDiskDevice["pool"] != newProfileRootDiskDevice["pool"]) {
+			// Check for any containers that have a local root disk device.
+			for i := 0; i < len(containers); i++ {
+				for _, v := range containers[i].LocalDevices() {
+					if v["type"] == "disk" && v["path"] == "/" && v["source"] == "" && v["pool"] != "" {
+						copy(containers[i:], containers[i+1:])
+						containers[len(containers)-1] = nil
+						containers = containers[:len(containers)-1]
+						i -= 1
+						break
+					}
+				}
+			}
+
+			// Check if another profile provides the same root disk device.
+			for i := 0; i < len(containers); i++ {
+				for _, p := range containers[i].Profiles() {
+					// Search through all of the containers
+					// profiles. Only skip the one we
+					// are about to change.
+					if p == profile.Name {
+						continue
+					}
+
+					_, pInfo, err := dbProfileGet(d.db, p)
+					if err != nil {
+						continue
+					}
+
+					for _, v := range pInfo.Devices {
+						// There's another profile that provides a root device on the same pool.
+						// So we can delete the container from the list.
+						if v["type"] == "disk" && v["path"] == "/" && v["source"] == "" && oldProfileRootDiskDevice["pool"] == v["pool"] {
+							copy(containers[i:], containers[i+1:])
+							containers[len(containers)-1] = nil
+							containers = containers[:len(containers)-1]
+							i -= 1
+							break
+						}
+					}
+				}
+			}
+			if len(containers) > 0 {
+				return BadRequest(fmt.Errorf("At least one container relies on this profiles' root disk device."))
+			}
+		}
+	}
```

Comment:
What we need to do during validation is:
 - Check that the expanded configuration doesn't contain two entries for any given "path" property.
 - With that validated, then only a single device can be the container's root device
 - As profiles override each other based on device names and container's overrides work the same way (masks the device based on its name), all we need to check is what's the source of the container's root disk entry. That's what my proposed code does by going through the sorted list of profiles in reverse order and checking for the first which defines that device's name. If the profile it finds is the one being modified, then that means that changes to the profile would affect the container.

What am I missing?